### PR TITLE
 Multicall CCIP-read gateway 

### PR DESCRIPTION
**Pull request by makoto**
Resolve https://github.com/ensdomains/offchain-resolver/issues/17

### CODE REVIEW

File: packages/batchclient/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,127 @@
+import { formatsByCoinType } from '@ensdomains/address-encoder';
```
Comment (Arachnid) [inline packages/batchclient/src/index.ts:1]: Can't all of this be replaced by a simple call to Ethers to resolve the name now?

File: packages/batchgateway/README.md [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+## Usage:
+You can run the gateway as a command line tool; in its default configuration it reads the data to serve from a JSON file specified on the command line:
+
+```
+yarn && yarn build
+yarn start --private-key 0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80 --data test.eth.json
+```
+
+`private-key` should be an Ethereum private key that will be used to sign messages. You should configure your resolver contract to expect messages to be signed using the corresponding address.
```
Comment (Arachnid) [inline packages/batchgateway/README.md:12]: Why is the batch gateway signing messages?

File: packages/batchgateway/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import { makeApp } from './server';
+import { Command } from 'commander';
+const program = new Command();
```
Comment (Arachnid) [inline packages/batchgateway/src/index.ts:3]:
```suggestion
import { Command } from 'commander';

const program = new Command();
```

File: packages/batchgateway/src/json.ts [hunk 4]
```diff
@@ -0,0 +1,76 @@
+import { Database } from './server';
```
Comment (Arachnid) [inline packages/batchgateway/src/json.ts:1]: This file should be unnecessary for the batch gateway.

File: packages/batchgateway/src/offchainexample.eth.json [hunk 5]
```diff
@@ -0,0 +1,25 @@
+{
```
Comment (Arachnid) [inline packages/batchgateway/src/offchainexample.eth.json:1]: This file should be unnecessary for the batch gateway.

File: packages/batchgateway/src/server.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+      type: 'query',
+      func: async ([data]: Result, request) => {
+        const sender = request.to;
+        let responses = await Promise.all(
+          data.map((d: any) => {
+            const url = d.urls[0];
+            const callData = d.callData;
+            const gatewayUrl = url
+              .replace('{sender}', sender)
+              .replace('{data}', callData);
```
Comment (Arachnid) [inline packages/batchgateway/src/server.ts:19]: If `{data}` isn't present, this should send a POST request, per EIP 3668.

File: packages/batchgateway/test.eth.json [hunk 7]
```diff
@@ -0,0 +1,18 @@
+{
```
Comment (Arachnid) [inline packages/batchgateway/test.eth.json:1]: This shouldn't be necessary for the batch gateway.

File: packages/batchgateway/src/utils.ts [hunk 8]
```diff
@@ -0,0 +1 @@
+export const ETH_COIN_TYPE = 60;
```
Comment (Arachnid) [inline packages/batchgateway/src/utils.ts:1]: This shouldn't be necessary for the batch gateway.

File: packages/batchgateway/test/e2e.test.ts [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+        return result;
+    }
+  }
+
+  detectNetwork(): Promise<ethers.providers.Network> {
+    return this.parent.detectNetwork();
+  }
+}
+
+const TEST_DB = {
```
Comment (Arachnid) [inline packages/batchgateway/test/e2e.test.ts:176]: None of this should be needed.

File: packages/batchgateway/test/e2e.test.ts [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+  return (
+    '0x' +
+    buf.reduce(
+      (output, elem) => output + ('0' + elem.toString(16)).slice(-2),
+      ''
+    )
+  );
+}
+
+describe('End to end test', () => {
```
Comment (Arachnid) [inline packages/batchgateway/test/e2e.test.ts:219]: This looks like it's testing the resolver gateway?

**Review by Arachnid — CHANGES_REQUESTED** [review]

**Comment by makoto** [comment]
Deprecated. The new PR is https://github.com/ensdomains/batch-gateway/pull/1

**Pull request closed without merging**